### DISCUSSION TITLE
Any feedback?

### DISCUSSION THREAD

**Discussion by Quafadas** [Quafadas/dedav4s#16] [General]
Has anyone used the library? Any observations, discussion of the paradigms etc?

**Comment by Stef500** [comment]
Hi, I've been trying to use your lib. It works great. Is there please a way to create a graph from a vega json file ?
I've been playing with Python and I got the json file. Now I'm trying to integrate it in scala with your package. Can you please help me with it? 
PS: I'm a noob in Scala but trying to learn ;-)

**Reply by Quafadas** [reply]
Hey @Stef500  - I'm happy to hear someone is giving this a go ... and I hope scala is treating you well :-). It sounds like you just want to plot "whatever is in some file"?

TL;DR: This sounded like a fairly common sort of desire, so I added it to my loosely maintained examples -  have a look here; 
https://github.com/Quafadas/viz-examples/blob/master/src/main/scala/Example_8.scala

I should move something like that class, into the library itself I think... 

Long, duller explanation; 

The core abstraction in the library is here; 
https://github.com/Quafadas/dedav4s/blob/main/core/shared/src/main/scala/viz/WithBaseSpec.scala

And it is, at heart very simple. We simple need to override the "baseSpec" property... with some valid (it doesn't check that) Vega spec.

The simplest possible strategy, might be to simple copy and paste some JSON into a case class.

```
case class SimplePlotOverride(override val mods: Seq[ujson.Value => Unit] = List()) extends WithBaseSpec:
  override lazy val baseSpec: ujson.Value = ujson.read(""" {"vega":"spec"} """)

```
In the example lined above, we've gone one step further, and read it from a file, then parsed it as json, using [oslib](https://github.com/com-lihaoyi/os-lib) and [ujson](https://www.lihaoyi.com/post/uJsonfastflexibleandintuitiveJSONforScala.html)

```
case class FromFilePlot(path: os.Path, override val mods: Seq[ujson.Value => Unit] = List()) extends WithBaseSpec(mods)  :
  override lazy val baseSpec: ujson.Value = ujson.read(os.read(path))
```

I hope that makes (some) sense? There's no real attempt at error handling (left to the reader :-)) here, but I hope this shows it can be  achieved simply. And by following that pattern we get all the other abstractions like plot targets, scalaJS support etc for free - just make sure you import the same stuff as on the "getting started page" of the docs.

Does that help? If you have any other questions - please feel free to open up another discussion and I'll happily answer them as best I can :-)! Happy plotting... and any other feedback - particularly on if it can be made "easier" to use, most welcome.

**Reply by Stef500** [reply]
Hi Simon, you just rock !!
I'm a convinced Pythonista trying to learn Scala, and it has been quite difficult to make the jump.
Your code did work very well, thanks a million. 
I might get back at you if I have more questions. 
Again, thanks ++ for your very detailed answer !
Best,
Stéphane

PS: as for the feedback, I think you should integrate this example into your documentation ;-)

**Reply by Quafadas** [reply]
> I'm a convinced Pythonista trying to learn Scala, and it has been quite difficult to make the jump.

Dunno what to say to that :-), Python is more popular than scala... and I guess there must be a good reason for that! If this library helps flatten the (undeniably rather steep) scala learning curve a little, then I shall be very proud!

I will say that as my domain got complex, and my code "bigger" and my time more constrained... I never looked back from scala, I do believe there are real benefits, but I think it's fair to say they might not be short term ones! FWIW - I'm pretty scared about "going back in" and making serious changes to programs I wrote in Python 4/5 years go. For whatever reason, fairly or unfairly, I don't have that fear in scala. Please note: I'm not a professional software engineer and I have nothing to sell you- a pro may have a totally different opinion... and python is clearly doing something right...

> I think you should integrate this example into your documentation ;-)

I'll put it one my list ;-). I actually haven't really done any maintenance on this library in the last year or so - I should probably try and touch it up a little...
